okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsFallbackStrategy.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2014 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.dnsoverhttps;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static okhttp3.dnsoverhttps.DnsOverHttps.isPrivateHost;
+
+/**
+ * A strategy for handling DNS lookup failures for DNS over HTTPS.
+ */
+public interface DnsFallbackStrategy {
```

Comment:
I don’t love this. I think this interface has a J2EE-feel . . . the amount of work it’s doing (a isPrivateHost() check) is dwarfed by the number of types and symbols and precedence orders.

Also weird to have DnsFallbackStrategy and DnsSource as things, rather than just having multiple implementations of Dns, and composing them.

Reply:
Yep, I wasn't sure.  I will refactor to that style (Dns decorator I mentioned above). Thanks for the advice.

Reply:
removed any notion of fallback for now, we aren't firefox, we don't need to *just* work for anything the user types in.  So supporting only the default happy path is reasonable for a mobile client, and probably better than a choose your own adventure of different strategies that may ultimately leak DNS traffic in response to a DOS attack.